x-pack/plugin/index-lifecycle/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleRunner.java

```diff
@@ -340,15 +342,33 @@ private static boolean canSetPolicy(StepKey currentStepKey, LifecyclePolicy curr
             if (currentPolicy.isActionSafe(currentStepKey)) {
                 return true;
             } else {
-                // Index is in the shrink action so fail it
-                // NORELEASE also need to check if the shrink action has changed between oldPolicy and newPolicy
-                return false;
+                // Index is in an unsafe action so fail it if the action has changed between oldPolicy and newPolicy
+                return isActionChanged(currentStepKey, currentPolicy, newPolicy) == false;
             }
         } else {
             // Index not previously managed by ILM so safe to change policy
             return true;
         }
     }
+    
+    private static boolean isActionChanged(StepKey stepKey, LifecyclePolicy currentPolicy, LifecyclePolicy newPolicy) {
+        LifecycleAction currentAction = getActionFromPolicy(currentPolicy, stepKey.getPhase(), stepKey.getAction());
+        LifecycleAction newAction = getActionFromPolicy(newPolicy, stepKey.getPhase(), stepKey.getAction());
+        if (newAction == null) {
+            return true;
+        } else {
+            return currentAction.equals(newAction) == false;
```

Comment:
since `getActionFromPolicy` can, theoretically, return `null`, would it be safer to switch these around to

```
newAction.equals(currentAction) == false
```

Reply:
right but if `currentAction` is `null` then we have a bug because we should never be in a situation where the current step doesn't exist in the current policy. If this does happen I would rather not silently fail returning a value. If this does not seem very friendly to you I can add a null check for `currentAction` which throws and `IllegalStateException` instead?

Reply:
I suppose that would make things super clear, but you're right, it would be a bug... so tossing a stacktrace vs. an illegalstateexception is almost the same thing 😄. all good